From 6dbc32e1b189076a1293d9655914b2c0742fd823 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marc-Andr=C3=A9=20Moreau?= <mamoreau@devolutions.net>
Date: Thu, 1 Feb 2024 13:53:38 -0500
Subject: [PATCH 6/8] fix non-windows build portability

---
 loginrec.c          | 2 +-
 monitor.c           | 1 +
 ssh-pkcs11-client.c | 2 ++
 3 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/loginrec.c b/loginrec.c
index 4f2149958..171a941c9 100644
--- a/loginrec.c
+++ b/loginrec.c
@@ -1033,7 +1033,7 @@ utmpx_perform_login(struct logininfo *li)
 		return (0);
 	}
 # else
-	if (!utmpx_write_direct(li, &ut)) {
+	if (!utmpx_write_direct(li, &utx)) {
 		logit("%s: utmp_write_direct() failed", __func__);
 		return (0);
 	}
diff --git a/monitor.c b/monitor.c
index f88fbb123..55974c50e 100644
--- a/monitor.c
+++ b/monitor.c
@@ -83,6 +83,7 @@
 #include "canohost.h"
 #include "log.h"
 #include "misc.h"
+#include "msg.h"
 #include "servconf.h"
 #include "monitor.h"
 #ifdef GSSAPI
diff --git a/ssh-pkcs11-client.c b/ssh-pkcs11-client.c
index 45cda44b5..fb7ab3923 100644
--- a/ssh-pkcs11-client.c
+++ b/ssh-pkcs11-client.c
@@ -509,10 +509,12 @@ pkcs11_start_helper(void)
 #endif
 	/* success */
 	debug3_f("started pid=%ld", (long)pid);
+#ifdef WINDOWS
 out:
 	if (client_token)
 		CloseHandle(client_token);
 	return r;
+#endif
 }
 
 int
-- 
2.25.1

